File: flappy_fly/game.py
"""Headless game with original geometry, deterministic seeds, and RGB frames."""

from dataclasses import dataclass
import math
import random

import numpy as np


@dataclass
class Pipe:
    x: float
    center: float
    passed: bool = False


class Flappy:
    width = 288
    height = 240
    ground = 232
    bird_x = 66.0
    radius = 5.0
    pipe_width = 30.0
    gap = 78.0
    spacing = 140.0
    scroll = 2.5
    gravity = 0.45
    impulse = -5.2

    def __init__(self, seed=0, max_frames=3600):
        if max_frames < 1:
            raise ValueError("max_frames must be positive")
        self.max_frames = max_frames
        self.reset(seed)

    def reset(self, seed=0):
        self.rng = random.Random(seed)
        self.seed = seed
        self.y = self.ground / 2
        self.velocity = 0.0
        self.frame = self.score = 0
        self.terminated = self.truncated = False
        self.pipes = [self._pipe(self.width + i * self.spacing) for i in range(3)]
        return self.observe()

    def _pipe(self, x):
        return Pipe(x, self.rng.uniform(self.gap / 2 + 22, self.ground - self.gap / 2 - 22))

    def observe(self):
        ahead = next(p for p in self.pipes if p.x + self.pipe_width >= self.bird_x - self.radius)
        return {
            "y": self.y,
            "velocity": self.velocity,
            "pipe_x": ahead.x,
            "gap_center": ahead.center,
            "score": self.score,
            "frame": self.frame,
        }

    def _hit_rectangle(self, left, top, right, bottom):
        x = min(max(self.bird_x, left), right)
        y = min(max(self.y, top), bottom)
        return (x - self.bird_x) ** 2 + (y - self.y) ** 2 <= self.radius ** 2

    def step(self, action):
        if self.terminated or self.truncated:
            raise RuntimeError("Reset the game after an episode ends")
        if not isinstance(action, (int, np.integer)) or action not in (0, 1):
            raise ValueError("Action must be 0 (wait) or 1 (flap)")
        if action:
            self.velocity = self.impulse
        self.velocity = min(self.velocity + self.gravity, 8.0)
        self.y += self.velocity
        self.frame += 1
        for pipe in self.pipes:
            pipe.x -= self.scroll
        self.terminated = self.y - self.radius <= 0 or self.y + self.radius >= self.ground
        for pipe in self.pipes:
            left, right = pipe.x, pipe.x + self.pipe_width
            self.terminated |= self._hit_rectangle(left, 0, right, pipe.center - self.gap / 2)
            self.terminated |= self._hit_rectangle(left, pipe.center + self.gap / 2, right, self.ground)
        reward = -1.0 if self.terminated else 0.01
        if not self.terminated:
            for pipe in self.pipes:
                if not pipe.passed and pipe.x + self.pipe_width < self.bird_x - self.radius:
                    pipe.passed = True
                    self.score += 1
                    reward += 1.0
        self.pipes = [p for p in self.pipes if p.x + self.pipe_width >= 0]
        while len(self.pipes) < 3:
            self.pipes.append(self._pipe(self.pipes[-1].x + self.spacing))
        self.truncated = self.frame >= self.max_frames and not self.terminated
        return self.observe(), reward, self.terminated, self.truncated

    def render(self):
        frame = np.empty((self.height, self.width, 3), dtype=np.uint8)
        frame[:] = (226, 242, 249)
        for pipe in self.pipes:
            left = max(0, math.floor(pipe.x))
            right = min(self.width, math.ceil(pipe.x + self.pipe_width))
            if right <= left:
                continue
            top = max(0, math.floor(pipe.center - self.gap / 2))
            bottom = min(self.ground, math.ceil(pipe.center + self.gap / 2))
            frame[:top, left:right] = (42, 142, 84)
            frame[bottom:self.ground, left:right] = (42, 142, 84)
        yy, xx = np.ogrid[:self.height, :self.width]
        bird = (xx - self.bird_x) ** 2 + (yy - self.y) ** 2 <= self.radius ** 2
        frame[bird] = (252, 188, 40)
        frame[self.ground:] = (150, 110, 65)
        return frame


def scene_generator_action(game):
    """Scripted stimulus generator only; never scored as a neural controller."""
    state = game.observe()
    return int(state["y"] + 5 * state["velocity"] > state["gap_center"] + 3)

File: docs/assets/game.mjs
import { LEVELS, SEEDS } from './levels.mjs';
export { SEEDS };

export class FlappyGame {
  static width = 288;
  static height = 240;
  static ground = 232;
  static birdX = 66;
  static radius = 5;
  static pipeWidth = 30;
  static gap = 78;
  static spacing = 140;

  constructor(seed = 19, maxFrames = 3600) {
    this.maxFrames = maxFrames;
    this.reset(seed);
  }

  reset(seed = 19) {
    if (!LEVELS[String(seed)]) throw new Error('Unknown browser level seed');
    this.seed = seed;
    this.level = LEVELS[String(seed)];
    this.nextPipe = 0;
    this.y = FlappyGame.ground / 2;
    this.velocity = this.frame = this.score = 0;
    this.terminated = this.truncated = false;
    this.pipes = Array.from({ length: 3 }, (_, i) => this.makePipe(288 + i * 140));
    return this.observe();
  }

  makePipe(x) {
    if (this.nextPipe >= this.level.length) throw new Error('Seeded level exhausted');
    return { x, center: this.level[this.nextPipe++], passed: false };
  }

  observe() {
    const pipe = this.pipes.find((p) => p.x + 30 >= 61);
    return {
      y: this.y,
      velocity: this.velocity,
      pipe_x: pipe.x,
      gap_center: pipe.center,
      score: this.score,
      frame: this.frame,
    };
  }

  hits(left, top, right, bottom) {
    const x = Math.min(Math.max(66, left), right);
    const y = Math.min(Math.max(this.y, top), bottom);
    return (x - 66) ** 2 + (y - this.y) ** 2 <= 25;
  }

  step(action) {
    if (this.terminated || this.truncated)
      throw new Error('Reset after an episode ends');
    if (action !== 0 && action !== 1) throw new Error('Action must be 0 or 1');
    if (action) this.velocity = -5.2;
    this.velocity = Math.min(this.velocity + 0.45, 8);
    this.y += this.velocity;
    this.frame += 1;
    for (const pipe of this.pipes) pipe.x -= 2.5;
    this.terminated = this.y - 5 <= 0 || this.y + 5 >= 232;
    for (const pipe of this.pipes) {
      this.terminated ||= this.hits(pipe.x, 0, pipe.x + 30, pipe.center - 39);
      this.terminated ||= this.hits(pipe.x, pipe.center + 39, pipe.x + 30, 232);
    }
    let reward = this.terminated ? -1 : 0.01;
    if (!this.terminated) {
      for (const pipe of this.pipes) {
        if (!pipe.passed && pipe.x + 30 < 61) {
          pipe.passed = true;
          this.score += 1;
          reward += 1;
        }
      }
    }
    this.pipes = this.pipes.filter((p) => p.x + 30 >= 0);
    while (this.pipes.length < 3)
      this.pipes.push(this.makePipe(this.pipes.at(-1).x + 140));
    this.truncated = this.frame >= this.maxFrames && !this.terminated;
    return {
      state: this.observe(),
      reward,
      terminated: this.terminated,
      truncated: this.truncated,
    };
  }
}

export function scriptedAction(game) {
  const state = game.observe();
  return Number(state.y > state.gap_center + 12);
}

export function drawGame(canvas, game, time = 0) {
  const rect = canvas.getBoundingClientRect();
  const ratio = Math.min(globalThis.devicePixelRatio || 1, 2);
  const width = Math.max(1, Math.round(rect.width * ratio));
  const height = Math.max(1, Math.round(rect.height * ratio));
  if (canvas.width !== width || canvas.height !== height) {
    canvas.width = width;
    canvas.height = height;
  }
  const ctx = canvas.getContext('2d');
  ctx.setTransform(width / 288, 0, 0, height / 240, 0, 0);
  const sky = ctx.createLinearGradient(0, 0, 0, 240);
  sky.addColorStop(0, '#10261f');
  sky.addColorStop(0.55, '#213d2a');
  sky.addColorStop(1, '#476344');
  ctx.fillStyle = sky;
  ctx.fillRect(0, 0, 288, 240);
  const light = ctx.createRadialGradient(220, 75, 0, 210, 95, 175);
  light.addColorStop(0, '#b2c88020');
  light.addColorStop(1, '#b2c88000');
  ctx.fillStyle = light;
  ctx.fillRect(0, 0, 288, 240);
  // Atmospheric scenery, independent of the neural simulation.
  for (let i = 0; i < 28; i++) {
    const x = ((((i * 73.7 - game.frame * 0.13) % 304) + 304) % 304) - 8;
    const y = ((i * 43.3) % 195) + 12;
    ctx.fillStyle = i % 3 === 0 ? '#d3e3a44d' : '#bdce9c24';
    ctx.fillRect(x, y, i % 4 === 0 ? 0.9 : 0.55, i % 4 === 0 ? 0.9 : 0.55);
  }
  for (let layer = 0; layer < 3; layer++) {
    ctx.beginPath();
    ctx.moveTo(0, 232);
    for (let x = 0; x <= 292; x += 4) {
      const offset = x + game.frame * (0.09 + layer * 0.1);
      const y =
        165 +
        layer * 23 +
        Math.sin(offset * 0.018 + layer * 2) * 13 +
        Math.sin(offset * 0.05 + layer) * 3;
      ctx.lineTo(x, y);
    }
    ctx.lineTo(292, 232);
    ctx.closePath();
    ctx.fillStyle = ['#436044', '#334f37', '#273f2e'][layer];
    ctx.fill();
  }
  ctx.strokeStyle = '#abc78312';
  ctx.lineWidth = 0.4;
  for (let x = 0; x < 288; x += 24) {
    ctx.beginPath();
    ctx.moveTo(x, 0);
    ctx.lineTo(x, 232);
    ctx.stroke();
  }
  for (const pipe of game.pipes) {
    for (const [y, h] of [
      [0, pipe.center - 39],
      [pipe.center + 39, 232 - pipe.center - 39],
    ]) {
      const metal = ctx.createLinearGradient(pipe.x, 0, pipe.x + 30, 0);
      metal.addColorStop(0, '#799765');
      metal.addColorStop(0.13, '#a8bd7b');
      metal.addColorStop(0.32, '#6a8a55');
      metal.addColorStop(0.8, '#344e32');
      metal.addColorStop(1, '#243f2b');
      ctx.fillStyle = '#0c1e1466';
      ctx.fillRect(pipe.x + 3, y + 3, 30, h);
      ctx.fillStyle = metal;
      ctx.fillRect(pipe.x, y, 30, h);
      ctx.fillStyle = '#d6e3a261';
      ctx.fillRect(pipe.x + 2, y, 0.7, h);
      ctx.fillStyle = '#223e2966';
      ctx.fillRect(pipe.x + 23, y, 1, h);
      ctx.fillStyle = '#182f2659';
      for (let rib = y + 12; rib < y + h - 7; rib += 14)
        ctx.fillRect(pipe.x + 4, rib, 22, 0.6);
      const capY = y === 0 ? h - 7 : y;
      ctx.fillStyle = '#182e23';
      ctx.fillRect(pipe.x, capY, 30, 7);
      ctx.fillStyle = '#829e62';
      ctx.fillRect(pipe.x + 0.5, capY + 0.5, 29, 5);
      ctx.fillStyle = '#d3e5a0';
      ctx.fillRect(pipe.x + 0.5, capY + (y === 0 ? 5 : 0.5), 29, 0.85);
      ctx.fillStyle = '#263d25';
      ctx.fillRect(pipe.x + 25, capY + 0.5, 4.5, 4.5);
      ctx.fillStyle = '#cbdc97';
      ctx.fillRect(pipe.x + 4, capY + 2, 1, 1);
      ctx.fillStyle = '#334b2e';
      ctx.fillRect(pipe.x + 21, capY + 2, 1, 1);
    }
  }
  ctx.fillStyle = '#17291c';
  ctx.fillRect(0, 232, 288, 8);
  ctx.fillStyle = '#98b06d';
  ctx.fillRect(0, 232, 288, 0.8);
  ctx.fillStyle = '#4b653b';
  for (let x = -((game.frame * 2.5) % 8); x < 288; x += 8) ctx.fillRect(x, 235, 4, 1.5);
  ctx.save();
  ctx.translate(66, game.y);
  ctx.rotate(Math.max(-0.3, Math.min(0.6, game.velocity * 0.065)));
  const wing = 0.55 + 0.45 * Math.sin(time * 0.045);
  ctx.fillStyle = '#e5f2cfd9';
  ctx.strokeStyle = '#b0c494';
  ctx.lineWidth = 0.45;
  for (const sign of [-1, 1]) {
    ctx.beginPath();
    ctx.ellipse(-1.5, sign * (4 + wing * 2), 6.5, 2.5, sign * 0.6, 0, Math.PI * 2);
    ctx.fill();
    ctx.stroke();
  }
  ctx.strokeStyle = '#182b20';
  ctx.lineWidth = 0.7;
  for (const offset of [-3, 0, 3]) {
    ctx.beginPath();
    ctx.moveTo(offset, 2);
    ctx.lineTo(offset - 2, 6);
    ctx.stroke();
  }
  ctx.fillStyle = '#d6af70';
  ctx.beginPath();
  ctx.ellipse(-2.5, 0, 5, 3.4, 0, 0, Math.PI * 2);
  ctx.fill();
  ctx.fillStyle = '#526446';
  ctx.fillRect(-5, -2.5, 1, 5);
  ctx.fillRect(-2.5, -3, 1, 6);
  ctx.fillStyle = '#59714e';
  ctx.beginPath();
  ctx.ellipse(1, 0, 3.5, 3.3, 0, 0, Math.PI * 2);
  ctx.fill();
  ctx.fillStyle = '#bfd298';
  ctx.beginPath();
  ctx.ellipse(0.4, -1.5, 2, 0.8, -0.1, 0, Math.PI * 2);
  ctx.fill();
  ctx.fillStyle = '#213326';
  ctx.beginPath();
  ctx.arc(4.2, -0.2, 3, 0, Math.PI * 2);
  ctx.fill();
  ctx.fillStyle = '#ff805c';
  ctx.beginPath();
  ctx.arc(5.2, -1, 2.2, 0, Math.PI * 2);
  ctx.fill();
  ctx.fillStyle = '#ffe5b7';
  ctx.fillRect(5, -2, 0.7, 0.7);
  ctx.restore();
}
